### ISSUE TITLE
Request: system tests for speech dictionaries

### ISSUE THREAD

**Issue by XLTechie** [nvaccess/nvda#14566]
### Is your feature request related to a problem? Please describe.

Because of some upcoming work I am contemplating on speech dictionaries, and an active current PR (#14557), I think it would be valuable to have some system tests for the basic correct performance of speech dictionaries. It would make testing more rapid, if the build system was already performing these tests; and would help provide a guide for future expectations.

### Describe the solution you'd like

System tests to be run against the speech dictionary UI.

### Tests proposed:

Initially, I propose these tests. This can always be expanded later.

*N.B. I imagine that speech replacement testing can be done with ui.message, but perhaps I am wrong about how the framework would work.*

* Open the panel to insert words into each dictionary: default, voice, temp.
    + In each dictionary panel, add the following patterns with replacements:
        - A case sensitive anywhere-match entry.
        - A case insensitive anywhere-match entry, which includes a backslash at the end of the replacement, per the error that #14557 is fixing.
        - A case sensitive word-match entry.
        - A case insensitive word-match entry.
        - A regular expression entry likely to be found in the wild, perhaps including group replacements.
    + Additionally, add another word-match entry to the default dictionary, to enable testing that two words in the same dictionary can be replaced.
* Speak one or more phrase(s) including the replacements, and be sure it/they is/are properly replaced. Note that multiple replacements should be triggered for a single spoken phrase, to test that multiple dictionaries/replacements can act on a single text chunk.
* Test the case sensitive words, by making sure that a phrase containing capitalized and non-capitalized versions of the pattern, is spoken with and without the replacement.
* After testing that one of the entries in the default dictionary is properly replaced, place an identical entry in the voice dictionary, and confirm that it is used instead of the default entry.
    + Do the same, but with an entry in the temp dictionary "covering over" the entries in voice and default.

I don't think we can test against a different voice, because of the method used to simulate speech. So further testing of voice dictionary handling is likely not possible currently, although my unfamiliarity with the testing framework might mean I'm wrong here.

### Describe alternatives you've considered

Continuing not to test the dictionaries. Maybe it isn't broken, so maybe we shouldn't fix it. But I think we should. :)

### Additional context

I could suggest some words and patterns, if whoever writes this doesn't want to use their own test data.

**Comment by seanbudd** [comment]
I think getting more unit test coverage would be better than system tests testing the the UX.
Our system tests are more slow and more fragile.
System tests are generally geared towards integration with other systems (e.g. Windows, Chrome, text editors).
System testing NVDA is mostly done to smoke to test important GUI functions that cannot be unit tested.

In this case, most of this logic could be tested internally with unit tests, without involving the NVDA GUI.
System testing on top of unit tests in this case would essentially be a test of the wxPython GUI.

**Comment by XLTechie** [comment]
@seanbudd I have updated this to request unit tests instead.
While I had some idea how to write system tests, and was sure I could figure it out from what we are already doing, I'm more than a bit under water with unit tests in NVDA.

I'm hoping `test_SpeechWithoutPauses.py` is a good example for how to monitor what would have been spoken after being mangled by the dictionaries
However, adding, removing, etc. dictionary entries, is GUI dependent at the moment as far as I can tell.
How would I go about simulating that with the unit testing framework? You can tell me to RTFM here, I'm just a bit uncertain about which M I should R.

**Comment by seanbudd** [comment]
>However, adding, removing, etc. dictionary entries, is GUI dependent at the moment as far as I can tell.
Yes the GUI action parts of this would be, but does it really need to be tested?
Other than general risks to the NVDA GUI and wxPython, are removing/adding entries at risk of breaking getting buggy?

>How would I go about simulating that with the unit testing framework? 
I suggest testing `speechDictHandler.SpeechDictEntry` and various classes and functions in `speechDictHandler` directly.
Testing `speechDictHandler.processText` may cover this at the highest level within unit tests.
You could simulate various dictionary setups by setting `speechDictHandler.dictionaries` during the unit test `setUp/tearDown`.
 `speechDictHandler.dictionaries` can be updated by appending `SpeechDictEntry`s.

**Comment by LeonarddeR** [comment]
@XLTechie Could you have a look at the unit tests in #19517 ? Do you think they are feasible to mark this issue as fixed?
